flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1729,13 +1794,34 @@ class ScalarExprLowering {
 
     auto call = builder.create<fir::CallOp>(loc, funcType.getResults(),
                                             funcSymbolAttr, operands);
+
+    if (caller.mustSaveResult())
+      builder.create<fir::SaveResultOp>(
+          loc, call.getResult(0), fir::getBase(allocatedResult.getValue()),
+          arrayResultShape, resultLengths);
+
     // Sync pointers and allocatables that may have been modified during the
     // call.
     for (const auto &mutableBox : mutableModifiedByCall)
       Fortran::lower::syncMutableBoxFromIRBox(builder, loc, mutableBox);
     // Handle case where result was passed as argument
-    if (caller.getPassedResult())
-      return resRef.getValue();
+
+    if (allocatedResult) {
+      return allocatedResult->match(
+          [&](const fir::MutableBoxValue &box) {
+            if (box.isAllocatable()) {
+              // 9.7.3.2 point 4. Finalize allocatables.
+              auto bldr = &builder;
```

Comment:
This may result in spurious crashes. `this` and all the member variables can be out of scope by the time the cleanup code is generated. (Recall that a single expression can switch from array to scalar and back several times.) Here bldr will be a pointer to `this->builder`. Use this instead:
```c++
  auto *bldr = &converter.getFirOpBuilder();
```

Reply:
Thanks for explaining.